src/healthcheck.c

```diff
@@ -0,0 +1,645 @@
+    if (!ht || !key) return false;
+    
+    unsigned int hash = hash_string(key) % ht->size;
+    struct hash_entry *entry = ht->buckets[hash];
+    
+    /* Check if key already exists */
+    while (entry) {
+        if (strcmp(entry->key, key) == 0) {
+            entry->value = value;
+            return true;
+        }
+        entry = entry->next;
+    }
+    
+    /* Create new entry */
+    entry = malloc(sizeof(struct hash_entry));
+    if (!entry) return false;
+    
+    entry->key = strdup(key);
+    if (!entry->key) {
+        free(entry);
+        return false;
+    }
+    
+    entry->value = value;
+    entry->next = ht->buckets[hash];
+    ht->buckets[hash] = entry;
+    ht->count++;
+    
+    return true;
+}
+
+/* Global healthcheck timers hash table */
+struct hash_table *active_healthcheck_timers = NULL;
+
+/* Check if systemd is available and running */
+bool healthcheck_is_systemd_available(void) {
+    /* Check if we're running under systemd by looking for NOTIFY_SOCKET */
+    const char *notify_socket = getenv("NOTIFY_SOCKET");
+    if (notify_socket != NULL && notify_socket[0] == '/') {
+        return true;
+    }
+    
+    /* Check if systemd is the init process (PID 1) */
+    FILE *fp = fopen("/proc/1/cmdline", "r");
+    if (fp) {
+        char cmdline[256];
+        if (fgets(cmdline, sizeof(cmdline), fp)) {
+            fclose(fp);
+            return strncmp(cmdline, "systemd", 7) == 0;
+        }
+        fclose(fp);
+    }
+    
+    return false;
+}
+
+/* Initialize healthcheck subsystem */
+bool healthcheck_init(void) {
+    if (active_healthcheck_timers != NULL) {
+        return true;
+    }
+    
+    active_healthcheck_timers = hash_table_new(16);
+    if (active_healthcheck_timers == NULL) {
+        return false;
+    }
+    
+    return true;
+}
+
+/* Cleanup healthcheck subsystem */
+void healthcheck_cleanup(void) {
+    if (active_healthcheck_timers != NULL) {
+        /* Free all timers first */
+        for (size_t i = 0; i < active_healthcheck_timers->size; i++) {
+            struct hash_entry *entry = active_healthcheck_timers->buckets[i];
+            while (entry) {
+                if (entry->value != NULL) {
+                    healthcheck_timer_free((healthcheck_timer_t*)entry->value);
+                    entry->value = NULL;  /* Prevent double-free */
+                }
+                entry = entry->next;
+            }
+        }
+        
+        /* Now free the hash table structure */
+        hash_table_free(active_healthcheck_timers);
+        active_healthcheck_timers = NULL;
+    }
+}
+
+/* Create a new healthcheck configuration */
+healthcheck_config_t *healthcheck_config_new(void) {
+    healthcheck_config_t *config = calloc(1, sizeof(healthcheck_config_t));
+    if (config == NULL) {
+        nwarn("Failed to allocate memory for healthcheck config");
+        return NULL;
+    }
+    
+    config->interval = 30;      /* Default 30 seconds */
```

Comment:
I'd want all of this passed explicitly. We shouldn't keep defaults in two places; Podman should pass every argument every time instead.